legion_core/src/world.rs

```diff
@@ -903,6 +1178,122 @@ impl Default for World {
+    fn clone_components(
+        &self,
+        src_world: &World,
+        src_component_storage: &ComponentStorage,
+        src_component_storage_indexes: core::ops::Range<ComponentIndex>,
+        src_type: ComponentTypeId,
+        src_entities: &[Entity],
+        dst_entities: &[Entity],
+        src_data: *const u8,
+        dst_data: *mut u8,
+        num_components: usize,
+    );
+}
+
+/// Used along with CloneImpl, allows receiving results from a clone_from or clone_from_single call
+pub trait CloneImplResult {
+    /// For every entity that is copied, this function will be called, passing the entity in the
+    /// source and destination worlds
+    fn add_result(&mut self, src_entity: Entity, dst_entity: Entity);
+}
+
+/// Used along with CloneImpl, allows specifying that certain entities in the receiving world should
+/// be replaced with entities from the source world
+pub trait EntityReplacePolicy<'s> {
+    /// This function must return all entities in the source world that will replace data in the
+    /// destination world
+    fn src_entities<'a>(&'s self) -> Box<dyn Iterator<Item = Entity> + 'a>
+    where
+        's: 'a;
+
+    /// This function must return all entities in the destination world that will be replaced
+    fn dst_entities<'a>(&'s self) -> Box<dyn Iterator<Item = Entity> + 'a>
+    where
+        's: 'a;
+
+    /// When called, if the provided source entity is intended to replace an entity in the
+    /// destination world, the function must return the entity in the destination world, or None
+    fn get_dst_entity(&self, src_entity: Entity) -> Option<Entity>;
+}
+
+/// Used to opt-out of receiving results from a clone_from or clone_from_single call
+pub struct NoneCloneImplResult;
+impl CloneImplResult for NoneCloneImplResult {
+    fn add_result(&mut self, _src_entity: Entity, _dst_entity: Entity) {
+        // do nothing
+    }
+}
+
+/// Used to opt-out of replacing entities during a clone_from or clone_from_single call
+pub struct NoneEntityReplacePolicy;
+impl<'s> EntityReplacePolicy<'s> for NoneEntityReplacePolicy {
+    fn src_entities<'a>(&self) -> Box<dyn Iterator<Item = Entity> + 'a>
+    where
+        's: 'a,
+    {
+        Box::new(std::iter::Empty::default())
+    }
+
+    fn dst_entities<'a>(&self) -> Box<dyn Iterator<Item = Entity> + 'a>
+    where
+        's: 'a,
+    {
+        Box::new(std::iter::Empty::default())
+    }
+
+    fn get_dst_entity(&self, _src_entity: Entity) -> Option<Entity> { None }
+}
+
+/// Default implementation of CloneImplResult that uses a hash map
+pub struct HashMapCloneImplResult<'m>(pub &'m mut HashMap<Entity, Entity>);
+
+impl<'m> CloneImplResult for HashMapCloneImplResult<'m> {
+    fn add_result(&mut self, src_entity: Entity, dst_entity: Entity) {
+        self.0.insert(src_entity, dst_entity);
+    }
+}
+
+/// Default implementation of EntityReplacePolicy that uses a hash map
+pub struct HashMapEntityReplacePolicy<'m>(pub &'m HashMap<Entity, Entity>);
```

Comment:
Isn't it better to make the default impl own the hashmap? 

```rust
        local.clone_from(
            &remote,
            &crate::create_copy_clone_impl(),
            &mut HashMapEntityReplacePolicy(&self.replace_mappings),
            &HashMapCloneImplResult(&mut self.result_mappings)
        );
```

It results in a strange API on my side like: 
```
            &mut HashMapEntityReplacePolicy(&self.replace_mappings),
            &HashMapCloneImplResult(&mut self.result_mappings)
```

Just a thought. I see the benefits as well, I will not have to give away my hashmap but just passing it over to the lib for a sec.

Reply:
Originally this function just took optional hash maps by reference (i.e. `Option<&HashMap<Entity, Entity>>`). It was cleaner but less flexible (and in particular, assumed downstream users were using hash maps with the default hashing algorithm). This is something @jaynus called out in the earlier review.

In a second pass, I experimented with implementing CloneImplResult and EntityReplacePolicy for hash map. This meant you could just pass a hash map by reference. I ended up not liking it because it felt error prone. The API took two HashMaps, both of which could be optional. It would be pretty easy to pass it into the wrong parameter or get them mixed up. Even if the end user got it right, if the API changed and we switched the order, downstream code would silently break.

So I changed it so that end users need to wrap the hash map in an explicit type (or provide their own impl for CloneImplResult and EntityReplacePolicy), which makes it a little harder to get it wrong, and avoids potential silent API breakage.

Another thing I don't like about this current API design is that a downstream user can't just pass in None if they don't care. (Even if we change the params to be Option, rust won't be able to infer the type inside the Option). The solution in this PR is to provide an implementation that does nothing.

In summary the alternatives I've considered are:
 * Just take HashMaps by reference (Simple but not flexible)
 * Implementing CloneImplResult and EntityReplacePolicy for hash map (Error prone)
 * Providing a simple wrapper that takes a ref to a hash map (Safe and flexible, but awkward API)
 * Just have a single trait instead of separate traits for cloning, receiving results, and specifying replacement (Removes the option of providing default implementations for two of the three traits)

I'm starting to think keeping separate traits for these 3 things is not worth the trouble. The end user will have to implement the CloneImpl trait no matter what, and merging CloneImplResult and EntityReplacePolicy are fairly trivial to implement.

Reply:
Indeed, hashmaps will be the most commonly used - if not only used - type here. I can't think of a better data structure for this use case. A trait will introduce extra complexity unless default implementations are provided by the library. A None and owning Hashmap implementations seems to be the most realistic. Though the downside of the owning hashmap implementation is that you take away the user control of that hashmap.

Reply:
Another pro of removing the trait objects is that the iter function as mentioned earlier doesn't have to be boxed.

Reply:
Removing the traits entirely means this would only work with HashMaps using the default harsher. It might be possible to make the function generic over the hashed type.

Reply:
Let's note the pros of having this abstraction:
- Very flexible. 
- Easier readable from the perspective of legion internal code. Common operations we do on those entities are hidden behind a name. 
- Less likely to break API in the future. 
- Easier to differentiate parameters.

Cons:
- Awkward API
- Can't pass in `None`
- Increased complexity for usage.


1. Implementation of traits for hashmap. As you said, and I had in mind as well, this seems like an obvious choice. However, we probably do not want this:

- API parameters are harder to differentiate and 
- We can't simply just-auto implement our trait for any hashmap out there. Users might not like auto implementation. 
- We would implement both traits for a hashmap, which means our hashmap can server as replace and result mapping. Which as well results in API-awkwardness. And more error-prone.

My idea is to let default implementations own the hashmap. This results in a slightly less awkward API. We can also implement a number of traits  for our default implementations to make usage easier (Deref, Clone, DerefMut). We ca

```rust
impl AsRef<HashMap<Entity, Entity>> for HashMapEntityReplacePolicy {
    fn as_ref(&self) -> &HashMap<Entity, Entity, RandomState> {
        &self.0
    }
}

impl AsMut<HashMap<Entity, Entity>> for HashMapEntityReplacePolicy {
    fn as_mut(&mut self) -> &mut HashMap<Entity, Entity, RandomState> {
        &mut self.0
    }
}

impl From<HashMap<Entity, Entity>> for HashMapEntityReplacePolicy {
    fn from(hashmap: HashMap<Entity, Entity, RandomState>) -> Self {
        HashMapEntityReplacePolicy(hashmap)
    }
}

impl Deref for HashMapEntityReplacePolicy {
    type Target = HashMap<Entity, Entity>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl DerefMut for HashMapEntityReplacePolicy {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}

```

This allows us to do:

```rust
let policy = HashMapEntityReplacePolicy::from(my_hashmap);
let clone = HashMapCloneImplResult::from(my_hashmap);

local.clone_from(&clone, &mut plolicy);

// access to underlying hashmap.
let hashmap_reference = &mut policy; // Deref and DerefMut
let hashmap_reference = policy.as_ref(); // AsRef and AsMut
```

Reply:
I think the default implementation should be as hands-off as possible, letting the user own their own data. We only need a reference for the lifetime of the function call - we shouldn't ask for any more than that.

(If an end user really wants to pass ownership, they can implement this themself.)